coreax/solvers/recombination.py

```diff
@@ -0,0 +1,570 @@
+        :code:`rcond is None`, it defaults to `floating point eps * max(n, d)`
+    :param tree_reduction_factor: The factor by which each tree reduction step reduces
+        the number of non-zero points; the remaining number of non-zero nodes, after
+        performing recombination, is equal to `n_nodes / tree_reduction_factor`;
+    """
+
+    rcond: Union[float, None] = None
+    tree_reduction_factor: int = 2
+
+    @override
+    def reduce(
+        self, dataset: Data, solver_state: None = None
+    ) -> tuple[Coresubset, None]:
+        nodes, weights = jtu.tree_leaves(dataset.normalize(preserve_zeros=True))
+        # Push the nodes forward through the test-functions \Phi^\prime.
+        push_forward_nodes = _push_forward(nodes, self.test_functions, augment=False)
+        n, m = push_forward_nodes.shape
+        # We don't apply the affine-augmentation test-function \phi_1 here, instead
+        # deferring it to `CaratheodoryRecombination.reduce`. Thus, we have to manually
+        # correct the value for `m`.
+        padding, count, depth = _prepare_tree(n, m + 1, self.tree_reduction_factor)
+        car_recomb_solver = CaratheodoryRecombination(rcond=self.rcond, mode="implicit")
+
+        def _tree_reduce(_, state):
+            """
+            Apply Tree-Based Caratheodory Recombination (Gaussian-Elimination).
+
+            Partitions the dataset into 'count' clusters of size 'n / count' and then
+            computes the cluster centroids. Caratheodory recombination is then performed
+            on these centroids (rather than on the full dataset), with every node in the
+            eliminated centroids' cluster being implicitly removed (given zero-weight).
+
+            There are 'tree_reduction_factor * m' clusters, with each step reducing the
+            number of remaining clusters down to 'm'. We can repeat the process until
+            each cluster contains, at most, a single non-zero weighted point (at this
+            point the recombination problem has been solved).
+            """
+            _weights, _indices = state
+            centroid_indices = jnp.argsort(_weights).reshape(count, -1, order="F")
+            centroid_nodes, centroid_weights = _centroid(
+                push_forward_nodes[_indices[centroid_indices]],
+                _weights[centroid_indices],
+            )
+            centroid_dataset = Data(centroid_nodes, centroid_weights)
+            centroid_coresubset, _ = car_recomb_solver.reduce(centroid_dataset)
+            coresubset_indices = centroid_coresubset.unweighted_indices
+            coresubset_weights = centroid_coresubset.coreset.weights
+            weight_update_indices = centroid_indices[coresubset_indices]
+            weight_update = coresubset_weights / centroid_weights[coresubset_indices]
+            updated_weights = _weights[weight_update_indices] * weight_update[..., None]
+            updated_indices = _indices[weight_update_indices.reshape(-1, order="F")]
```

Comment:
More comments describing the maths here would be helpful

Reply:
I thought the code in this section was reasonably self-explanatory (probably beacuse I've been looking at it for a while) so would you be able to provide any specifics that you would like expanding on?

Reply:
If there are any equation references in the paper you can refer to that would be great